File: 15.hs
import Data.List
import Data.List.Split

main = do
  inp <- getContents
  solution inp

step (x:xs) = case elemIndex x xs of
  Just i -> (1 + i):x:xs
  Nothing -> 0:x:xs 
 
play n l = if n == length l 
  then head l
  else play n $ step l

solution inp = do
  let nums = map (read :: String -> Int) (splitOn "," inp)
  print $ play 2020 (reverse nums)
  -- takes forever. never finishes?
  --print $ play 30000000 (reverse nums)

File: 15.js
let inp = [0,6,1,7,2,19,20];
let N = 3e7;
let m = {};
let v = inp.pop();
inp.forEach((n, i) => m[n] = 1 + i);
for (let n = 1 + inp.length; n < N; n++) {
  let i = m[v] || n;
  m[v] = n;
  v = n - i;
}
console.log(v);
